File: optimizers/Adadelta.py
from .Optimizer import Optimizer, np


class Adadelta(Optimizer):
    """
    Adadelta — адаптивный метод на экспоненциальных средних квадратов градиентов и квадратов обновлений.

    Первоисточник: Zeiler M.D. "ADADELTA: An Adaptive Learning Rate Method". arXiv preprint arXiv:1212.5701, 2012.
    https://arxiv.org/abs/1212.5701
    Реализация следует алгоритму из документации PyTorch (в отличие от статьи Зейлера,
    обновление дополнительно умножается на скорость обучения lr):
    https://docs.pytorch.org/docs/stable/generated/torch.optim.Adadelta.html
    """

    param_descriptions = {
        "lr": "скорость обучения — дополнительный масштаб обновления (есть в PyTorch, в статье Зейлера отсутствует)",
        "rho": "коэффициент затухания накопителей квадратов градиентов и квадратов обновлений",
        "eps": "малая добавка под корнями: численная защита и стартовый масштаб шага, пока накопители пусты",
    }

    acc1: np.ndarray
    acc2: np.ndarray

    def __init__(
        self, initial_x: np.ndarray, function, lr: float = 0.01, rho: float = 0.975, eps: float = 1e-02
    ) -> None:
        params = dict(lr=lr, rho=rho, eps=eps)
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)
        self.acc1 = self.params["rho"] * self.acc1 + (1 - self.params["rho"]) * gradient**2
        update = gradient * np.sqrt(self.acc2 + self.params["eps"]) / np.sqrt(self.acc1 + self.params["eps"])
        self.acc2 = self.params["rho"] * self.acc2 + (1 - self.params["rho"]) * update**2
        next_x = self.x - self.params["lr"] * update
        return self.move_next(next_x)

    def reset(self) -> None:
        super().reset()
        self.acc1 = np.zeros([2])
        self.acc2 = np.zeros([2])

File: web/src/shared/lib/optimization-engine/optimizers/adadelta.ts
import type { OptimizerDescriptor } from "./types";
import { sub2, zero2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient } from "@shared/lib/optimization-engine/functions";

export const adadeltaOptimizer: OptimizerDescriptor = {
  name: "Adadelta",
  params: {
    lr: {
      default: 0.01,
      description: "скорость обучения — дополнительный масштаб обновления (есть в PyTorch, в статье Зейлера отсутствует)",
    },
    rho: { default: 0.975, description: "коэффициент затухания накопителей квадратов градиентов и квадратов обновлений" },
    eps: {
      default: 1e-2,
      description: "малая добавка под корнями: численная защита и стартовый масштаб шага, пока накопители пусты",
    },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let acc1: Vec2 = zero2();
    let acc2: Vec2 = zero2();
    return {
      get x() { return x; },
      params,
      next() {
        const grad = gradient(fn, x[0], x[1]);
        acc1 = [
          params.rho * acc1[0] + (1 - params.rho) * grad[0] ** 2,
          params.rho * acc1[1] + (1 - params.rho) * grad[1] ** 2,
        ];
        const update: Vec2 = [
          (grad[0] * Math.sqrt(acc2[0] + params.eps)) / Math.sqrt(acc1[0] + params.eps),
          (grad[1] * Math.sqrt(acc2[1] + params.eps)) / Math.sqrt(acc1[1] + params.eps),
        ];
        acc2 = [
          params.rho * acc2[0] + (1 - params.rho) * update[0] ** 2,
          params.rho * acc2[1] + (1 - params.rho) * update[1] ** 2,
        ];
        x = sub2(x, [params.lr * update[0], params.lr * update[1]]);
        return { x, value: fn(x[0], x[1]) };
      },
      reset() {
        x = initialX;
        acc1 = zero2();
        acc2 = zero2();
      },
    };
  },
};
